Scripts/BuildPhases/LintAppLocalizedStringsUsage.swift

```diff
@@ -0,0 +1,324 @@
+            PBXProject.self,
+            PBXGroup.self,
+            PBXFileReference.self,
+            PBXNativeTarget.self,
+            PBXSourcesBuildPhase.self,
+            PBXBuildFile.self
+        ]
+
+        init(from decoder: Decoder) throws {
+            // Try to decode each known types in turn, until we find one that succeeds decoding — by having the expected `isa` field value.
+            for objectType in Self.knownTypes {
+                if let object = try? objectType.init(from: decoder) as PBXObject {
+                    self.wrappedValue = object
+                    return
+                }
+            }
+            self.wrappedValue = try UnknownPBXObject(from: decoder) as PBXObject // Fallback
+        }
+        var debugDescription: String { String(describing: wrappedValue) }
+    }
+}
+
+
+
+// MARK: - Lint method
+
+/// The outcome of running our lint logic on a file
+enum LintResult { case ok, skipped, violationsFound([(line: Int, col: Int)]) }
+
+/// Lint a given file for usages of `NSLocalizedString` instead of `AppLocalizedString`
+func lint(fileAt url: URL, target: String) throws -> LintResult {
+    guard ["m", "swift"].contains(url.pathExtension) else { return .skipped }
+    let content = try String(contentsOf: url)
+    var lineNo = 0
+    var violations: [(line: Int, col: Int)] = []
+    content.enumerateLines { line, _ in
+        lineNo += 1
+        guard line.range(of: "\\s*//", options: .regularExpression) == nil else { return } // Skip commented lines
+        guard let range = line.range(of: "NSLocalizedString") else { return }
+
+        let colNo = line.distance(from: line.startIndex, to: range.lowerBound)
+        let message = "Use `AppLocalizedString` instead of `NSLocalizedString` in source files that are used in the `\(target)` extension target. See paNNhX-nP-p2 for more info."
+        print("\(url.path):\(lineNo):\(colNo): error: \(message)")
+        violations.append((lineNo, colNo))
+    }
+    return violations.isEmpty ? .ok : .violationsFound(violations)
+}
+
+
+
+// MARK: - Main (Script Code entry point)
+
+// 1st arg = project path
+let args = CommandLine.arguments.dropFirst()
+guard let projectPath = args.first, !projectPath.isEmpty else { print("You must provide the path to the xcodeproj as first argument."); exit(1) }
+let project = try Xcodeproj(path: projectPath)
+
+// 2nd arg (optional) = name of target to lint
+let targetsToLint: [Xcodeproj.PBXNativeTarget]
+if let targetName = args.dropFirst().first, !targetName.isEmpty {
+    targetsToLint = project.nativeTargets.filter { $0.name == targetName }
+} else {
+    targetsToLint = project.nativeTargets.filter { $0.productType == .appExtension }
+}
+
+// Lint each requested target
+var violationsFound = 0
+for target in targetsToLint {
+    let files: [Xcodeproj.PBXBuildFile] = project.buildFiles(for: target)
+    print("Linting the Build Files for \(target.name):")
+    for file in files {
```

Comment:
I don't image performance is an issue for this script, but I feel like [task group](https://docs.swift.org/swift-book/LanguageGuide/Concurrency.html#ID642) is perfect for this use case here, which potentially can speed up this script a little bit. I'm not sure if it's feasible to use the structured concurrency in a standalone script though...

Reply:
Yeah interesting thought, but I'm not sure it's worth it for this context.

That being said… I've been eager to play with structured concurrency and Task Groups (I haven't had the occasion to do much Swift in a while… and in fact working on this Swift implementation of the linter was a great excuse to get back the pleasure of doing some Swift again) so maybe that would be a good exclude to look at this in an future PR to improve on it if I find the time… 🙂

Reply:
No worries, I don't think there is really much practical gain in using it here, other than trying some new shiny tech.

Reply:
For the record: I actually tried using `TaskGroup` and it worked but:

 - Because you can't call `async` code from a non-async context:
    - I had to wrap that "main" part of the script inside a `Task { … }` block
    - … and then I had to use a `DispatchSemaphore` (that I'd `signal` at the end of the `Task { … }`) so that I could block the main thread and wait until all tasks were done. It works, but felt a bit dirty to have to rely on a semaphore… 🤷 
 - It messed up the order of the lines I `print`-ed in the `stdout`, obviously because I parsed my files and thus wrote my logs in an async order now.
    - That wouldn't have been too strange if those logs were just a big flat list, but since I print a tree structure, i.e. list of targets then list of files scanned in that target, that messed things up
    - Sure I could have localized the parallelization per target only, and/or actually used the _structured_ part of structured concurrency to create some task dependencies… but I didn't go that far
    - (In a similar vein I thought about using `AsyncSequence` to read the url.lines` when linting each file, but didn't spend enough time on this to get to that)
 - And ultimately, at least according to my crude estimation of invoking my script using `time <myscript>` a couple of time and averaging the results for both before and after… it appeared that using structured concurrency took… more time overall!! I guess it's the cost of creating threads under the hood for so many small tasks that don't overcome the benefit we'd otherwise have with the parallelisation.

That was still a really nice and fun thing to try, and good excuse to _finally_ play with `TaskGroup`, `async`/`await`, and structured concurrency… so thanks for the suggestion, I had fun anyway 😄

Reply:
> it appeared that using structured concurrency took… more time overall!!

🤣 this is awesome. Now I learned the switching thread context may indeed have negative impact on performance. 

Thanks for trying it and sharing the learnings! 👍